src/node_trace_events.cc

```diff
@@ -0,0 +1,166 @@
+}
+
+static const char* GetCategoryGroup(Environment* env,
+                                    const Local<Value>& categoryValue) {
+  if (!categoryValue->IsString()) {
+    env->ThrowTypeError("trace_events category must be a string.");
+    return nullptr;
+  }
+
+  Utf8Value category(env->isolate(), categoryValue);
+  // If the value already exists in the set, insertion.first will point
+  // to the existing value. Thus, this will maintain a long lived pointer
+  // to the category c-string.
+  auto insertion = categoryGroups.insert(category.out());
+
+  // The returned insertion is a pair whose first item is the object that was
+  // inserted or that blocked the insertion and second item is a boolean
+  // indicating whether it was inserted.
+  return insertion.first->c_str();
+}
+
+static inline char GetPhase(Environment* env, const Local<Value>& arg) {
+  if (!arg->IsNumber()) {
+    env->ThrowTypeError("trace_events type must be a number.");
+    return 0;
+  }
+  Local<Context> context = env->isolate()->GetCurrentContext();
+  return static_cast<char>(arg->Int32Value(context).ToChecked());
+}
+
+static inline int64_t GetId(Environment* env, const Local<Value>& arg, int64_t& id) {
+  if (arg->IsUndefined() || arg->IsNull()) {
+    id = 0;
+    return true;
+  }
+  else if (arg->IsNumber()) {
+    // This value may be zero, if no id is used
+    Local<Context> context = env->isolate()->GetCurrentContext();
+    id = arg->IntegerValue(context).ToChecked();
+    return true;
+  }
+  else {
+    env->ThrowTypeError("trace_events id must be a non-negative number.");
+    return false;
+  }
+}
+
+static void Emit(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+
+  // Args: [type, category, name, id, argName, argValue]
+  CHECK_GE(args.Length(), 3);
+
+  // Check the category group first, to avoid doing more work if it's not
+  // enabled.
+  const char* category_group = GetCategoryGroup(env, args[1]);
+  const uint8_t* category_group_enabled = GetCategoryGroupEnabled(category_group);
+  if (*category_group_enabled == 0) return;
+
+  char phase = GetPhase(env, args[0]);
+  if (phase == 0) return;
+
+  if (!args[2]->IsString()) {
+    env->ThrowTypeError("trace_events name must be a string.");
+    return;
+  }
+  Utf8Value nameValue(env->isolate(), args[2]);
+  const char* name = nameValue.out();
+
+  int64_t id = 0;
+  if (args.Length() >= 4 && !GetId(env, args[3], id)) return;
+
+  // TODO(@AndreasMadsen): Two extra arguments are not supported.
+  // TODO(@AndreasMadsen): String values are not supported.
+  int32_t num_args = 0;
+  const char* arg_names[1];
+  uint8_t arg_types[1];
+  uint64_t arg_values[1];
+
+  if (args.Length() < 6 || (args[4]->IsUndefined() && args[5]->IsUndefined())) {
+    num_args = 0;
+  }
+  else if (args[4]->IsString() && args[5]->IsNumber()) {
+    num_args = 1;
+    Local<Context> context = env->isolate()->GetCurrentContext();
```

Comment:
You can just use `env->context()`, and if you want to have `context` be a local variable I’d put it somewhere near the top (where `env` is declared as well)